### PR TITLE
Allow getting factories from component Host

### PR DESCRIPTION
**Pull request by tigrannajaryan**
Added GetFactory function to Host interface. This allows components
to fetch factories and create components dynamically. The functionality
will be used by receiver creator that creates and destroys other receivers
as it discovers relevant external endpoints from which these receivers
can get data.

### CODE REVIEW

**Comment by jrcamp** [comment]
LGTM

File: component/component.go [hunk 1]
```diff
@@ -40,6 +51,18 @@ type Host interface {
...
+	// This allows components to create other components. For example:
+	//   func (r MyReceiver) Start(host component.Host) error {
+	//     apacheFactory := host.GetFactory(KindReceiver,"apache").(component.ReceiverFactory)
+	//     receiver, err := apacheFactory.CreateMetricsReceiver(...)
+	//     ...
+	//   }
+	// GetFactory can be called by the component anytime after Start() begins and
+	// until Shutdown() is called. Note that the component is responsible for destroying
+	// other components that it creates.
+	GetFactory(kind Kind, componentType string) Factory
```
Comment (bogdandrutu) [inline component/component.go:65]: Do you think we should have a `type ComponentType string` to make sure we can extend that in the future if needed? Maybe just an issue now.

File: testbed/testbed/load_generator.go [hunk 2]
```diff
@@ -134,7 +134,7 @@ func (lg *LoadGenerator) generate() {
 	// Indicate that generation is done at the end
 	defer lg.stopWait.Done()
 
-	if lg.options.DataItemsPerSecond <= 0 {
+	if lg.options.DataItemsPerSecond == 0 {
```
Comment (bogdandrutu) [inline testbed/testbed/load_generator.go:137]: why this change? Maybe a separate PR?

File: component/component.go [hunk 3]
```diff
@@ -40,6 +51,18 @@ type Host interface {
 	// Context returns a context provided by the host to be used on the component
 	// operations.
 	Context() context.Context
```
Comment (bogdandrutu) [inline component/component.go:53]: Unrelated to this PR, but I am not sure I understand this Context function.

File: testbed/testbed/load_generator.go (hunk 2 above)
Comment (tigrannajaryan) [reply testbed/testbed/load_generator.go:137]: One of linters or checkers was complaining, don't remember which one.

File: component/component.go (hunk 3 above)
Comment (tigrannajaryan) [reply component/component.go:53]: @pjanotti may be able to explain, I don't remember.

File: component/component.go (hunk 1 above)
Comment (tigrannajaryan) [reply component/component.go:65]: I will look into this separately since it is declared a string in many other places and changing it may break other code that imports us.
Issue added https://github.com/open-telemetry/opentelemetry-collector/issues/762

File: component/component.go (hunk 3 above)
Comment (bogdandrutu) [reply component/component.go:53]: @pjanotti can you take this as an action item to update the comment and explain what is that function for and how is it used.

File: testbed/testbed/load_generator.go (hunk 2 above)
Comment (bogdandrutu) [reply testbed/testbed/load_generator.go:137]: Ah I see DataItemsPerSecond is uint so cannot be negative :)

**Comment by tigrannajaryan** [comment]
@bogdandrutu can you change this repo settings to allow merging out-of-date PRs? There is a small chance to breaking things this way but it is annoying to require rebasing and rebuilding PRs because master keeps moving forward.

File: component/component.go (hunk 3 above)
Comment (pjanotti) [reply component/component.go:53]: @bogdandrutu I'm unsure but this looks like a leftover from when some components needed a common context to start their operations and this was passed via the `Host`. It will make more sense to remove it from here and add it to Start. I will check the appropriate alternative.
